homeassistant/components/google_drive/config_flow.py

```diff
@@ -0,0 +1,98 @@
+import logging
+from typing import Any
+
+from aiohttp.client_exceptions import ClientError
+
+from homeassistant.config_entries import SOURCE_REAUTH, ConfigFlowResult
+from homeassistant.const import CONF_ACCESS_TOKEN, CONF_TOKEN
+from homeassistant.helpers import config_entry_oauth2_flow
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .api import create_headers
+from .const import (
+    DEFAULT_NAME,
+    DOMAIN,
+    DRIVE_API_FILES,
+    DRIVE_FOLDER_URL_PREFIX,
+    OAUTH2_SCOPES,
+)
+
+
+class OAuth2FlowHandler(
+    config_entry_oauth2_flow.AbstractOAuth2FlowHandler, domain=DOMAIN
+):
+    """Config flow to handle Google Drive OAuth2 authentication."""
+
+    DOMAIN = DOMAIN
+
+    @property
+    def logger(self) -> logging.Logger:
+        """Return logger."""
+        return logging.getLogger(__name__)
+
+    @property
+    def extra_authorize_data(self) -> dict[str, Any]:
+        """Extra data that needs to be appended to the authorize url."""
+        return {
+            "scope": " ".join(OAUTH2_SCOPES),
+            # Add params to ensure we get back a refresh token
+            "access_type": "offline",
+            "prompt": "consent",
+        }
+
+    async def async_step_reauth(
+        self, entry_data: Mapping[str, Any]
+    ) -> ConfigFlowResult:
+        """Perform reauth upon an API authentication error."""
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Confirm reauth dialog."""
+        if user_input is None:
+            return self.async_show_form(step_id="reauth_confirm")
+        return await self.async_step_user()
+
+    async def async_oauth_create_entry(self, data: dict[str, Any]) -> ConfigFlowResult:
+        """Create an entry for the flow, or update existing entry."""
+
+        session = async_get_clientsession(self.hass)
+        headers = create_headers(data[CONF_TOKEN][CONF_ACCESS_TOKEN])
+
+        if self.source == SOURCE_REAUTH:
+            return self.async_update_reload_and_abort(
+                self._get_reauth_entry(), data=data
+            )
+
+        try:
+            resp = await session.post(
+                DRIVE_API_FILES,
+                params={"fields": "id"},
+                json={
+                    "name": "Home Assistant",
+                    "mimeType": "application/vnd.google-apps.folder",
+                    "properties": {
+                        "ha": "root",
+                    },
+                },
+                headers=headers,
+            )
+            resp.raise_for_status()
+            res = await resp.json()
+        except ClientError as err:
+            self.logger.error("Error creating folder: %s", str(err))
+            return self.async_abort(reason="create_folder_failure")
+        folder_id = res["id"]
+        await self.async_set_unique_id(folder_id)
```

Comment:
So why do we pick the folder_id over for example an user id?

Reply:
But they are unique on the system right?

Reply:
What do you mean?

Reply:
The unique id we set here is only unique to this system right?

Or how would it work if I set up 2 instances with the same google drive account?

Reply:
The folder_id is the unique id of the created Home Assistant folder in Google Drive. If you set up another HA instance with the same Google Drive account, a separate Home Assistant folder with its own unique id will be created. Note, Google Drive allows having multiple files and folders with the same name in the same folder.

FYI, the Google Sheets integration, from which I copied a lot of code, including tests, is using the sheet file id as the unique id. BTW, the main difference of config_flow.py, api.py, and __init__.py between the 2 integrations is that I moved the check whether the folder exists from the reauth flow to the setup to satisfy the test-before-setup requirement.

Reply:
> Note, Google Drive allows having multiple files and folders with the same name in the same folder.

Currently the integration is single_config_entry, which is fine for now, but if in the future it would open up, I don't think we want to allow to add multiple accounts which all have their own folder. I would rather block it at user level then. WDYT? (I am not saying we should change that but I do like to push back to see why we do the things we do)

Reply:
> I don't think we want to allow to add multiple accounts which all have their own folder. 
Why not? In fact I initially I didn't have single_config_entry because I wanted to allow connecting to different Google Drive accounts in case someone wanted extra redundancy with multiple backups in different Google accounts. But it seems you can only have one application credentials per integration. Without the single_config_entry I remember I wasn't getting the prompt to enter new application credentials. It immediately opened the page on the Google side to authorize access. That's why I added the single_config_entry.

Reply:
I removed the single_config_entry, changed unique_id to be the email address and changed the folder and backup files to include the HA instance uuid.